### ISSUE TITLE
[JENKINS-34268] Lock multiple resources using the Pipeline lock step

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/lockable-resources-plugin#840]
<p>The current implementation of Pipeline <tt>lock</tt> step allows to block a single resource.</p>

<p>It should be extended to cover all the functionality of the plugin (applicable to non-freestyle jobs) such as blocking resources by label or request a lock for N resources.</p>

<p>The DSL must be something like this:</p>

<pre>
lock (resources: [<span class="code-quote">'resource1'</span>, <span class="code-quote">'resource2'</span>]) {
  ... execution block ...
}
</pre>

<p>or</p>

<pre>
lock (label: <span class="code-quote">'my-resources'</span>) {
  ... execution block ...
}
</pre>

<p>The behavior of the <tt>label</tt> parameter would be equivalent to:</p>

<pre>
lock (resources: [<span class="code-quote">'resource3'</span>, <span class="code-quote">'resource4'</span>]) { <span class="code-comment">// <span class="code-keyword">if</span> both resource3 and resource4 are labeled as <span class="code-quote">'my-resources'</span>
</span>  ... execution block ...
}
</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amuniz">amuniz</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268" target="_blank">Lock multiple resources using the Pipeline lock step</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amuniz">amuniz</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: lockable-resources-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-12-19T18:12:11+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 33
<li><b>imported</b>: 20251217-081411
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The current implementation of Pipeline <tt>lock</tt> step allows to block a single resource.</p>

<p>It should be extended to cover all the functionality of the plugin (applicable to non-freestyle jobs) such as blocking resources by label or request a lock for N resources.</p>

<p>The DSL must be something like this:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock (resources: [<span class="code-quote">'resource1'</span>, <span class="code-quote">'resource2'</span>]) {
  ... execution block ...
}
</pre>
</div></div>

<p>or</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock (label: <span class="code-quote">'my-resources'</span>) {
  ... execution block ...
}
</pre>
</div></div>

<p>The behavior of the <tt>label</tt> parameter would be equivalent to:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock (resources: [<span class="code-quote">'resource3'</span>, <span class="code-quote">'resource4'</span>]) { <span class="code-comment">// <span class="code-keyword">if</span> both resource3 and resource4 are labeled as <span class="code-quote">'my-resources'</span>
</span>  ... execution block ...
}
</pre>
</div></div></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [PR](https://github.com/jenkinsci/lockable-resources-plugin/pull/42)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aeon512:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=254517&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254517">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Besides the actual label functionality, also the possibility to request a certain amount of resources from this label would be essential.</p>

<p>I would personally prefer something like</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock(label: <span class="code-quote">'fooBar'</span>, quantity: 2) {
  <span class="code-comment">// whatever
</span>}
</pre>
</div></div>

<p>Actually this is what keeps us from using pipelines.</p>
</pre>
</details>

</ul>
</details>

<p>Besides the actual label functionality, also the possibility to request a certain amount of resources from this label would be essential.</p>

<p>I would personally prefer something like</p>


<pre>
lock(label: <span class="code-quote">'fooBar'</span>, quantity: 2) {
  <span class="code-comment">// whatever
</span>}
</pre>

<p>Actually this is what keeps us from using pipelines.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> amuniz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=254523&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254523">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aeon512" class="user-hover" rel="aeon512">aeon512</a> Could you explain a real use case where the <tt>quantity</tt> configuration makes sense please? It's not clear to me why one would want to acquire 2 resources but do not caring about which ones specifically.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aeon512" class="user-hover" rel="aeon512">aeon512</a> Could you explain a real use case where the <tt>quantity</tt> configuration makes sense please? It's not clear to me why one would want to acquire 2 resources but do not caring about which ones specifically.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aeon512:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=254527&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254527">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><b>Background:</b>
We dynamically create virtual machines which acts as jenkins slave (using the <tt>swarm-plugin</tt>) to perform the actual build steps. Thereby, if an error occurs we can archive the whole VM such that a developer can investigate and solve the problem. Using linked snapshots the whole purpose of creating such a new VM takes simply seconds and works very well.</p>

<p><b>Problem solved by lockable resources:</b>
Since the overall amount of memory and cores on the host machine is limited, we used the lockable resources plugin to manage the distribution of memory (slices) and cpu cores.</p>

<p>For example, let's say we have configured 16 resources (Memory Slice 1, Memory Slice 2, ..., Memory Slice 3), each representing e.g. 4 GB Memory. that is a total of 64 GB being available for virtual machines. When a job starts and wants to create a VM with e.g. 16 GB virtual memory, the job simply tries to request 4 slices ( <tt>=quantity</tt> ) of 4 GB memory</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock(label: <span class="code-quote">'memory'</span>, quantity: 4) {
   <span class="code-comment">// actual build steps
</span>}
</pre>
</div></div>
<p>If this succeeds there is enough memory available, and the job can continue. If not, the job blocks until the necessary amount of memory is available again (since other jobs have finished and shut down their VMs).</p>

<p>The same principle applies to CPU (cores) as well.</p>

<p>Does this explanation help?</p>
</pre>
</details>

</ul>
</details>

<p><b>Background:</b><br/>
We dynamically create virtual machines which acts as jenkins slave (using the <tt>swarm-plugin</tt>) to perform the actual build steps. Thereby, if an error occurs we can archive the whole VM such that a developer can investigate and solve the problem. Using linked snapshots the whole purpose of creating such a new VM takes simply seconds and works very well.</p>

<p><b>Problem solved by lockable resources:</b><br/>
Since the overall amount of memory and cores on the host machine is limited, we used the lockable resources plugin to manage the distribution of memory (slices) and cpu cores.</p>

<p>For example, let's say we have configured 16 resources (Memory Slice 1, Memory Slice 2, ..., Memory Slice 3), each representing e.g. 4 GB Memory. that is a total of 64 GB being available for virtual machines. When a job starts and wants to create a VM with e.g. 16 GB virtual memory, the job simply tries to request 4 slices ( <tt>=quantity</tt> ) of 4 GB memory</p>

<pre>
lock(label: <span class="code-quote">'memory'</span>, quantity: 4) {
   <span class="code-comment">// actual build steps
</span>}
</pre>
<p>If this succeeds there is enough memory available, and the job can continue. If not, the job blocks until the necessary amount of memory is available again (since other jobs have finished and shut down their VMs).</p>

<p>The same principle applies to CPU (cores) as well.</p>

<p>Does this explanation help?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amuniz.png" title="amuniz's avatar" /> amuniz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=254533&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254533">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for take the time to explain!</p>

<p>Makes sense, although nothing is preventing the steps inside the <tt>lock</tt> block to consume more CPU/RAM than virtually acquired, right?</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for take the time to explain!</p>

<p>Makes sense, although nothing is preventing the steps inside the <tt>lock</tt> block to consume more CPU/RAM than virtually acquired, right?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aeon512:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=254536&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-254536">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes and no <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
<b>No:</b> Once the VM is configured to a certain amount of CPU/RAM, that amount is fixed and no process inside the VM can use more. If a process inside such a VM want's to consume more memory it simply receives an OutOfMemory exception. 
<b>Yes:</b> However, in theory, you can <tt>lock</tt> a different amount than you configure the VM.</p>

<p>However, in our pipeline scripts we define a variable, e.g. <tt>required_memory_slices</tt> and use this variable to specify the necessary quantity as well as use this variable when setting up the virtual machine. Hence, in this case, it is guaranteed that we lock the same amount that we configure the VM to use, and hence no process can consume more CPU/RAM.</p>
</pre>
</details>

</ul>
</details>

<p>Yes and no <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
<b>No:</b> Once the VM is configured to a certain amount of CPU/RAM, that amount is fixed and no process inside the VM can use more. If a process inside such a VM want's to consume more memory it simply receives an OutOfMemory exception. <br/>
<b>Yes:</b> However, in theory, you can <tt>lock</tt> a different amount than you configure the VM.</p>

<p>However, in our pipeline scripts we define a variable, e.g. <tt>required_memory_slices</tt> and use this variable to specify the necessary quantity as well as use this variable when setting up the virtual machine. Hence, in this case, it is guaranteed that we lock the same amount that we configure the VM to use, and hence no process can consume more CPU/RAM.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nwegner:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=271578&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271578">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>May be the locked resource might be return via given variable as requested in <a href="https://issues.jenkins.io/browse/JENKINS-30269" title="Add workflow support for resource locking" class="issue-link" data-issue-key="JENKINS-30269"><del>JENKINS-30269</del></a>. This would be a similar way as if configured thru a job config.</p>
</pre>
</details>

</ul>
</details>

<p>May be the locked resource might be return via given variable as requested in <a href="https://issue-redirect.jenkins.io/issue/30269" title="Add workflow support for resource locking" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/30269">JENKINS-30269</a>"><del>JENKINS-30269</del></a>. This would be a similar way as if configured thru a job config.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fkykko:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=279785&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-279785">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Will PR36 be merged/released soon? This would resource allocation of some old ssh builders that we would like to do from pipelines. I there anything I can do to help? Is there a test build that we could run?</p>
</pre>
</details>

</ul>
</details>

<p>Will PR36 be merged/released soon? This would resource allocation of some old ssh builders that we would like to do from pipelines. I there anything I can do to help? Is there a test build that we could run?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>fkykko:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=280961&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-280961">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amuniz" class="user-hover" rel="amuniz">amuniz</a>
Hi! Nice to see this merged.
Looking at the merged content, I can't seem to find any pipeline syntax for actually knowing what resource was locked.
Specifically the proposed syntax from PR36 above:  ", variable: 'MY_VAR'" is missing.
Our use case is that we wan't to specify a label containing node names for legacy builders that can't run the slave.jar (java7) anymore. We wan't to lock a resources (host name) from he label and then ssh to that machine any execute some build steps. This works with the FreestyleJob functionality of #Reserved resources variable name". Any thought of adding this to pipeline?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=amuniz" class="user-hover" rel="amuniz">amuniz</a><br/>
Hi! Nice to see this merged.<br/>
Looking at the merged content, I can't seem to find any pipeline syntax for actually knowing what resource was locked.<br/>
Specifically the proposed syntax from PR36 above:  ", variable: 'MY_VAR'" is missing.<br/>
Our use case is that we wan't to specify a label containing node names for legacy builders that can't run the slave.jar (java7) anymore. We wan't to lock a resources (host name) from he label and then ssh to that machine any execute some build steps. This works with the FreestyleJob functionality of #Reserved resources variable name". Any thought of adding this to pipeline?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>glance:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=281694&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281694">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I did a really ugly workaround for the missing resourceVariable :</p>

<p>lock(label: 'LABEL', quantity: 1) {
    echo org.jenkins.plugins.lockableresources.LockableResourcesManager.class.get().getResourcesFromBuild(currentBuild.getRawBuild())<span class="error">&#91;0&#93;</span>.getName()
}</p>

<p>That will get you which resource you locked.</p>
</pre>
</details>

</ul>
</details>

<p>I did a really ugly workaround for the missing resourceVariable :</p>

<p>lock(label: 'LABEL', quantity: 1) {<br/>
    echo org.jenkins.plugins.lockableresources.LockableResourcesManager.class.get().getResourcesFromBuild(currentBuild.getRawBuild())<span class="error">&#91;0&#93;</span>.getName()<br/>
}</p>

<p>That will get you which resource you locked.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mireksz:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=310010&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310010">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glance" class="user-hover" rel="glance">glance</a></p>
<blockquote><p>I did a really ugly workaround for the missing resourceVariable :</p></blockquote>

<p>Did you happen to find any way of getting the resource that's locked in the current lock closure? Your workaround always returns the name of first locked resource. I'm trying to run label-based locks in parallel, so I cannot simply use getResourcesFromBuild(...)<span class="error">&#91;0&#93;</span>, but must find out the exact resource locked within current scope. Any ideas?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=glance" class="user-hover" rel="glance">glance</a></p>
<blockquote><p>I did a really ugly workaround for the missing resourceVariable :</p></blockquote>

<p>Did you happen to find any way of getting the resource that's locked in the current lock closure? Your workaround always returns the name of first locked resource. I'm trying to run label-based locks in parallel, so I cannot simply use getResourcesFromBuild(...)<span class="error">&#91;0&#93;</span>, but must find out the exact resource locked within current scope. Any ideas?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>glance:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=310012&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-310012">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The hack grew quite a bit over the following days, and ended up as:</p>

<p><a href="https://gist.github.com/glance-/aaa3c037757895798d4e1be5134bb843" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/glance-/aaa3c037757895798d4e1be5134bb843</a></p>

<p> </p>

<p>I removed the need for extensive whitelisting by writing it as a pipeline library which later could be imported into the pipeline dsl by:
@Library("PipelineHelpers")
import PipelineHelpers.LockableResourcesHelper</p>

<p> </p>

<p>We never use different locked resources in parallel so I have never needed to figure out if you can map locked resource to which block that locked it. If you ever do, please post the solution here for others to find.</p>
</pre>
</details>

</ul>
</details>

<p>The hack grew quite a bit over the following days, and ended up as:</p>

<p><a href="https://gist.github.com/glance-/aaa3c037757895798d4e1be5134bb843" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/glance-/aaa3c037757895798d4e1be5134bb843</a></p>

<p> </p>

<p>I removed the need for extensive whitelisting by writing it as a pipeline library which later could be imported into the pipeline dsl by:<br/>
@​Library("PipelineHelpers")<br/>
import PipelineHelpers.LockableResourcesHelper</p>

<p> </p>

<p>We never use different locked resources in parallel so I have never needed to figure out if you can map locked resource to which block that locked it. If you ever do, please post the solution here for others to find.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>armb:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=311734&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311734">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We use locked resources in parallel, so the getResourcesFromBuild workaround didn't work for us.</p>

<p><a href="https://github.com/jenkinsci/lockable-resources-plugin/pull/50" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/lockable-resources-plugin/pull/50</a> seems to work. (I haven't tried pull/49, but it appears to be a similar approach to the same fix.)</p>
</pre>
</details>

</ul>
</details>

<p>We use locked resources in parallel, so the getResourcesFromBuild workaround didn't work for us.</p>

<p><a href="https://github.com/jenkinsci/lockable-resources-plugin/pull/50" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/lockable-resources-plugin/pull/50</a> seems to work. (I haven't tried pull/49, but it appears to be a similar approach to the same fix.)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>tgr:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-34268?focusedId=381224&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381224">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=chrisdickinsonnio" class="user-hover" rel="chrisdickinsonnio">chrisdickinsonnio</a> Can you open a new ticket instead of commenting on a closed one (where nobody will ever see your comments, I stumbled on it by accident) if you have issues with the plugin? Anyway, locking of multiple resources is implemented since 2.3. The syntax (as generated by the snippet generator) is:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">lock(extra: [[resource: <span class="code-quote">'b'</span>], [label: <span class="code-quote">'c'</span>, quantity: 1]], resource: <span class="code-quote">'a'</span>) {
    <span class="code-comment">// some block
</span>}
</pre>
</div></div>
<p> </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=chrisdickinsonnio" class="user-hover" rel="chrisdickinsonnio">chrisdickinsonnio</a> Can you open a new ticket instead of commenting on a closed one (where nobody will ever see your comments, I stumbled on it by accident) if you have issues with the plugin? Anyway, locking of multiple resources is implemented since 2.3. The syntax (as generated by the snippet generator) is:</p>

<pre>
lock(extra: [[resource: <span class="code-quote">'b'</span>], [label: <span class="code-quote">'c'</span>, quantity: 1]], resource: <span class="code-quote">'a'</span>) {
    <span class="code-comment">// some block
</span>}
</pre>
<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/38165">JENKINS-38165</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/30269">JENKINS-30269</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/34273">JENKINS-34273</a>]</i>
